### PR TITLE
[jog_arm] Fix initial end effector transform jump

### PR DESCRIPTION
**Pull request by AndyZe**
- Fix a bug where the end effector transform jumped initially, when the first jog command is received.

- Reduce duplicate code between joint jogging and Cartesian jogging.

- Remove the joint velocity low-pass filters. I justify this because joint velocities are derived from joint positions (which do get filtered).

- This required some re-ordering of the algorithm. I think the flow makes a lot more sense now.

### CODE REVIEW

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 1]
```diff
@@ -111,9 +108,28 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
...
+    incoming_joints_ = shared_variables.joints;
+    mutex.unlock();
+
+    kinematic_state_->setVariableValues(joint_state_);
+
+    // Always update the end-effector transform in case the getCommandFrameTransform() method is being used
+    // Get the transform from MoveIt planning frame to jogging command frame
+    // We solve (planning_frame -> base -> robot_link_command_frame)
+    // by computing (base->planning_frame)^-1 * (base->robot_link_command_frame)
+    tf_moveit_to_cmd_frame_ = kinematic_state_->getGlobalLinkTransform(parameters_.planning_frame).inverse() *
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Does this really fix the issue or is it just much more unlikely to occur? To me it seems like there is still the possibility of a race condition when getCommandFrameTransform() is called before we have an updated joint state. Also, wouldn't it be better to compute `tf_moveit_to_cmd_frame` inside getCommandFrameTransform() in order to reduce redundant computations? I don't see the transform being used here anymore. Maybe we should additionally let `getCommandFrameTransform()` return bool or add a blocking function `waitForValidJointState()`.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 2]
```diff
@@ -124,25 +140,19 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
...
+    // Always update the joints and end-effector transform for 2 reasons:
+    // 1) in case the getCommandFrameTransform() method is being used
+    // 2) so the low-pass filters are up to date and don't cause a jump
     mutex.lock();
     incoming_joints_ = shared_variables.joints;
     mutex.unlock();
-
+    while (!updateJoints() && ros::ok() && !stop_requested_)
+    {
+      ros::Duration(WHILE_LOOP_WAIT).sleep();
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Maybe unrelated to this PR, but is there a reason we are not using `ros::Rate`?

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 3]
```diff
@@ -153,91 +163,96 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
...
-      mutex.lock();
-      // If everything normal, share the new traj to be published
-      if (valid_nonzero_command)
+      // Prioritize cartesian jogging above joint jogging
+      if (!zero_cartesian_cmd_flag)
       {
-        shared_variables.outgoing_command = outgoing_command_;
-        shared_variables.ok_to_publish = true;
+        mutex.lock();
+        cartesian_deltas = shared_variables.command_deltas;
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: All these mutexes make the code unnecessarily hard to read. We should add thread safe accessor functions to shared_variables

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 4]
```diff
@@ -153,91 +163,96 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
...
-          ++zero_velocity_count;
+        suddenHalt(outgoing_command_);
+        zero_cartesian_cmd_flag = true;
+        zero_joint_cmd_flag = true;
+      }
+
+      bool valid_nonzero_command = !zero_cartesian_cmd_flag || !zero_joint_cmd_flag;
+
+      // Send the newest target joints
+      if (!outgoing_command_.joint_names.empty())
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Are there ever no joint names specified in the message at this point?

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 5]
```diff
@@ -153,91 +163,96 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
     shared_variables.tf_moveit_to_cmd_frame = tf_moveit_to_cmd_frame_;
     mutex.unlock();
 
-    // Initialize the position filters to initial robot joints
-    while (!updateJoints() && ros::ok() && !stop_requested_)
+    // Only do jogging calculations if the robot should move, for efficiency
+    if (stop_requested_)
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Am I reading this wrong or does `stop_requested_` now only mean that the positions are reset, but the loop is kept alive? That's not what `stop_requested_` is supposed to do, we need an API to actually stop the loop as well. Stopping should be done by resetting the velocities in the command.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 1 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Line 94 ensures joint_state_ is updated before this, so no race condition.

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Yeah, I guess it does make sense to calculated `tf_moveit_to_cmd_frame` inside the getter method.

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: However, I did notice that `tf_planning_to_cmd_frame` was redundant with `tf_moveit_to_cmd_frame_` and could be deleted

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I added an `isInitialized()` function and `std::atomic<bool> is_initialized_` like you added previously for `stop_requested_`

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 2 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Not that I know of.  ;)  Will update it

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I found out there is a reason. The default sleep rate is defined in jog_arm_data.h:

`static const double WHILE_LOOP_WAIT = 0.001;`

If I change that to `static ros::Rate DEFAULT_SLEEP_RATE(1000)'` the error is:

Cannot use ros::Time::now() before the first NodeHandle has been created or ros::start() has been called

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 3 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: OK, but in a different PR, right?

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 4 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: not that i know of, will try deleting it

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: deleting it _seems_ fine...

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 3 above)
Comment (jliukkonen) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: It should go a long way if you refactored JogArmShared into a class with mutexed setters and getters. That should ensure no race conditions can occur. Also note that some of the variables can just be converted into atomics.

Another thing I would recommend, is to use `std::lock_guard<foo>`. The thing about locking and unlocking is, that eventually you forget to unlock and there will be problems. Lock guard always unlocks the mutex at the end of scope so you can't really forget anything. You can also do this: 
```
std::lock_guard<std::mutex> lock(mutex)
{
  // Do things here
} // automatic unlocking.
```

In other words,, you set a temporary scope just to guide the lock_guard.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 6]
```diff
@@ -153,91 +163,96 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
...
+          shared_variables.ok_to_publish = true;
+        }
+        mutex.unlock();
+
+        // Store last zero-velocity message flag to prevent superfluous warnings.
+        // Cartesian and joint commands must both be zero.
+        if (zero_cartesian_cmd_flag && zero_joint_cmd_flag)
+        {
+          // Avoid overflow
+          if (zero_velocity_count < INT_MAX)
```
Comment (jliukkonen) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: You can use `std::numeric_limits<int>::max()`. That will also warn about incompatible types. With macros I'm not sure whether e.g., signed / unsigned comparisons are caught or not.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 5 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: The robot will not move if `stop_requested_` is true but the node will continue to make the end_effector_transform available (for closed loop controllers). It does look like the usage of `stop_requested_` needs updating in some places, though.

I think the way to exit from the loop is to destroy the object. Otherwise, we want it to continue to spin and provide the end effector transform (unless getEndEffectorTransform() should not be provided by this node.)

Over a year ago, the design decision was that you shouldn't need to publish zeros to stop motion. Because
a) ROS1 messages can be dropped
b) if you're spamming zeros to the controller, you shut down other ways of moving the robot. People may have multiple motion control modes, not just jogging. E.g. trajectory control and joint jogging possibly could operate with the same ros_control controller

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I guess this confusion was due to the naming of the variable. How about renaming:

`stop_requested_ -> exit_requested_`

To signify that it's not just meant to halt, but to exit.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 6 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: done

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 5 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: done

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 3 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I'll work on this in a followup PR. This one's already pretty big

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 1 above)
Comment (henningkayser) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I thought it would be enough to just access `shared_variables.joints` for getting the latest state. Anyway, this doesn't really belong into this PR and I'm fine with the current solution

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 2 above)
Comment (henningkayser) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Yes, ros::Rate can't be static, you'll have to initialize it before the loop (or inside constructor). A big advantage is that ros::Rate will run the loop in the rate you specified while `ros::Duration::sleep()` adds a constant sleep additional to the computation time.

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: ok, done

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 3 above)
Comment (jliukkonen) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I'm going to approve this so we can move on to "fix mutexes" state :)
